### ISSUE TITLE
Maria:  Headless cms, 3 weeks

### ISSUE THREAD

**Issue by MMikhailova** [lab-brussels-1/home#401]
# headless cms
## Learning goals

- 🥚 Understanding what a content management system is, and why people use it 
- 🥚 Given an existing CMS installation fill in and publish a new piece of content
- 🥚 Perform a simple http call via postman to retrieve a piece of information.
- 🐣 Extend and adapt an existing CMS configuration
- 🐣 Perform a http call with filtering via Postman
- 🐣 Perform a http call with "include" via Postman
- 🐣 Perform a http POST call to create new data
- 🐣 Perform a http PUT call to update existing data
- 🐣 Configuring endpoint permissions (Authentication vs Authorization)
- 🐥 Design a data model from scratch based on user requirements

### Suggested study

- <https://www.youtube.com/watch?v=vcopLqUq594>
- <https://www.youtube.com/watch?v=6FnwAbd2SDY>
- <https://strapi.io/blog/understanding-and-using-relations-in-strapi>

### Reference Material during exercises

- [Rest API documentation](https://docs.strapi.io/developer-docs/latest/developer-resources/database-apis-reference/rest-api.html)

### Chapter 1 - Introduction

In this first chapter we are going to take a look at what a Content Management System is and why a company might want to use it.

### Chapter 2 - HTTP Calls

In this chapter we are going to take a look at how to set up endpoints and retrieve the data we are interested in. Starting from basic examples we will slowly work our way into more and more difficult scenarios using Postman.

### Chapter 3 - Data Model Design

In this chapter we will put ourselves in the role of analyst and set out to design the data model required to fulfill the client's needs.

### Chapter 4 - Authentication vs Authorization

In this chapter we will take a look at the difference between Authentication and Authorization. We will see how you can secure your endpoints by using role based user accounts.

## Deliverables

Throughout this module you will choose one of the cases you analysed during Chapter 3 and design and implement the model in Strapi. 

Using [strapi-template-repo](https://github.com/HackYourFutureBelgium/template-strapi-vanilla),You will  build a minimal frontend application that connects to your CMS and that uses proper authentication to access certain routes.

This is a group project.

- Fork the the template repo to **Lab organization**. 

- When connecting to strapi, **one person from the team** has to create the admin name and password and share it with the rest of the team.

Checkout [this](https://youtu.be/h9vETeRiulY?t=39) link on how to get started with with strapi

- [ ] [README repository](https://github.com/_/_)
- [ ]  live README](<https://github.com/_>)
- I have practiced:
  - [ ] pulling and pushing
  - [ ] formatting my code: `npm run format`
  - [ ] checking my spelling: `npm run spell-check`
  - [ ] linting my code: `npm run lint:ls` _and_ `npm run lint:md`

---- END CHECK-IN TEMPLATE ---->

**Comment by MMikhailova** [comment]
> > Why don't we use Web version of Strapi?
> 
> there is no web version at the moment, and even when there will be, it will not be free.
> 

Now you see that I was confused with notions..
By saying web version I meant localhost:1337/admin in the browser. 
I must have written the note before I had studied the topic well enough 😊 

Anyway now It's clearer for me. Thank you

**Comment by MMikhailova** [comment]
Week 2

I pushed my progress [here](https://github.com/MMikhailova/chat-app).

## I Need Help With:

I have implemented Chat app data model in Strapi and managed to retrieve the data via Postman and JavaScript.
I am currently working on displaying/modifying data via HTML&JS. I managed to display Users and Channels, sending messages (via POST to Message table). I suppose to be able to send a message from different users I need first to make sign in form. This step is not clear for me yet. Hope we will go through authentication and "sign in" subjects on Sunday.


## What went well?

- Chat app model implementation
- Fetching data from server via complex JS queries using filters, sort, populate and etc.
- Displaying data via HTML&CSS page 

I also enjoy playing in [fantastic-moxie](https://fantastic-moxie-8bcf10.netlify.app/) by @yoshimalaise 👍 

## What went less well?

I got sick and almost lost my voice again and I feel bad when I have to refuse some help requests from classmates:(

## Lessons Learned

Work life balance is important.

## Sunday Prep Work

Keep working on chat - app.

**Comment by danielhalasz** [comment]
> I suppose to be able to send a message from different users I need first to make sign in form. This step is not clear for me yet. Hope we will go through authentication and "sign in" subjects on Sunday.

yes, that is the plan

> I got sick and almost lost my voice again

oooooh, very sad to hear that! will you be able to attend in person on Sunday?

**Comment by danielhalasz** [comment]
CONGRATULATIONS once again for the hard work you did with the chat app!

**Comment by MMikhailova** [comment]
## Me

### What was not clear, where did you get stuck?

First I got stuck with the difference between registration and login in Strapi, but thanks to @danielhalasz explication it became crystal clear:) 
The only thing that I couldn't get so far is how to upload an image to avatar field during the registration from JS file.
I tried POST method to `http://localhost:1337/api/users?populate=avatar` with 
```js 
{
    data:{
   avatar: {
       url:"/upload/hyf.jpeg"
   }
    }
}
```
But it doesn't work. I had to upload img from Strapi content manager.

### What was clear, what did you master?
This module was one of the most exciting for me. 
Yoshi is a great teacher!

- I have learnt data modelling in Strapi
- Managed to implement chat-app data model with HTML/CSS and JavaScript
- Finally configured endpoint permissions (Authentication) 

### Where can you still use some help?

I can't understand how can I share my chat-app code. Will it work in Git page if I create one for the repo?

I would really like to finish my chat-app and it would be great if somebody could give a feedback on my code.

### Where can you help others moving forwards?
I was happy to help all classmates who asked me for this in this module and I plan to keep sharing.

---

## The Course

### What can there be more of?
Everything was enough.

### What can there be less of?

I don't know))

### What material were most helpful (from HYF or elsewhere)?
As usual, I focused only on suggested studies and it was enough to understand the basics.

### What HYF material was least helpful?
Strapi documentation.

### Any suggestions for future classes?

THANK YOU!

**Comment by danielhalasz** [comment]
> But it doesn't work. I had to upload img from Strapi content manager.

indeed, as far as I know you can not store media files directly in a SQL database, but instead upload them to a separate service such as Cloudinary or use the content manager.

**Comment by danielhalasz** [comment]
> I can't understand how can I share my chat-app code. Will it work in Git page if I create one for the repo?

Github pages only works for the frontend, the backend has to be hosted on a separate server. In the individual projects module with Yoshi, we will look into how to deploy a full stack application. We are actually currently looking into the options, as the service we used to use at HYF, Heroku, is no longer free.
